custom-schemes.md

```diff
@@ -2,40 +2,54 @@
 
 Existing specifications like [OpenID4VP](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#name-wallet-invocation) and [ISO 18013-7](https://www.iso.org/standard/82772.html) currently define the use of custom URL schemes (eg. `openid4vp:` and `mdoc:`) to enable verifier websites to invoke holder (wallet) applications. This approach works in some scenarios, but has a number of problems we aim to solve by designing a browser API which can replace the use of custom URL schemes in these specifications. This DRAFT document aims to outline the questions and concerns with the use of custom URL schemes for identity presentment in order to provide more detail on potential benefits of a dedicated browser API.
 
-1. Can wallets reliably determine their invoker?
+## Can wallets reliably determine their invoker?
 
-   - Android and iOS offer no facility to determine the web origin which triggered a navigation to a custom URL scheme. 
+   - Android and iOS offer no facility to determine the web origin which triggered a navigation to a custom URL scheme.
    - This seems to increase the opportunity for attacks such as replay attacks, man-in-the-middle attacks and phishing. It also seems to limit the opportunity for abuse detection and prevention.
    - A browser API can securely pass the requesting origin on to the wallet so that the wallet can know with confidence what origin the user was on when the request was initiated.
+   
+## How are wallet apps supposed to pass info back to the invoking page?
 
-2. Can wallets limit requests to secure contexts?
+Does the page give the wallet a URL to navigate back to as part of the custom scheme URL it uses to launch the wallet?
 
-   - If a user uses a website via http (not https) and it invokes a custom scheme, it's possible that an attacker with access to the network could eavesdrop on or even modify the request (such as by replacing the verifier's public key with their own). 
-   - A browser API could take the position that it will function only within the context of a secure context (https).
+If so, that's a problem because it won't go back to the same tab. Apps don't have the ability to return information to the same page that invoked them after a custom scheme invocation; they can just open a new page in a new tab.
 
-3. What is the user experience when multiple wallet apps are installed?
+## Can wallets limit requests to secure contexts?
+
+   - If a user uses a website via http (not https) and it invokes a custom scheme, it's possible that an attacker with access to the network could eavesdrop on or even modify the request (such as by replacing the verifier's public key with their own).
+   - A browser API could (and almost certainly should) take the position that it will function only within the context of a secure context (https).
```

Comment:
That's more of a problem of the website and ecosystems can take precautions against this (rules, checks for HTTPS redirections, HSTS, etc.).

Reply:
Leaving security up to "ecosystems" tends to result in more security holes than we'd like to see. The mere fact that you can invoke an OID4VP flow via an HTTP website is a security issue, and no amount of lecturing web developers will prevent that from happening. A browser API would help guide developers towards only operating in HTTPS mode, universally, which is a good thing.

Reply:
I'm OK with relying on the ecosystem but only if we're confident the incentives are aligned with the agency. I think it's issuers who have the most incentive for security and so they're going to want to certify wallets. Most wallets will probably want to ensure they're invoked securely (even if only to earn the trust of the credential issuers). But verifiers may not have as much incentive, they're more incentivized to make adoption as easy as possible. So if wallets can't impose security requirements on verifiers then I think we have a break in the accountability chain and should expect security for users to suffer. Having the platform in the middle seems to fix this because the platform is incentivized to earn the trust of wallets/issuers (otherwise they'll go around the platform API), and so will impose security constraints on verifiers which increase trust of the wallets and issuers.